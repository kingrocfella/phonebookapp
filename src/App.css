.display-end {
  display: flex;
  justify-content: flex-end;
}

.user-avatar {
  background-color: blue;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}
