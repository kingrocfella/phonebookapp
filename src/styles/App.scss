@import "./variables";

html,
body {
  background-color: $secondary-color;
}

.display-end {
  display: flex;
  justify-content: flex-end;
}

.display-center-end {
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.primary-button {
  background-color: $primary-color;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.block-button {
  background-color: $primary-color;
  color: white;
}

.delete-button {
  background-color: $danger;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.user-avatar {
  background-color: $primary-color;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}

.display-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.display-vertical-center {
  display: flex;
  align-items: center;
}

.hovered-row {
  background-color: $cardboxshadow;
}

.contact-table {
  border-spacing: 0rem;
  border-collapse: separate;
}

.space-between {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.invisible {
  visibility: hidden;
}
