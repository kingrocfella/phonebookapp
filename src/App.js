import { useState } from "react";
import Modal from "./components/Modal";
import AddContact from "./components/addContact";
import ContactList from "./components/contactList";
import ContactDetails from "./components/contactDetails";
import { ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "./App.css";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState([]);

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={deleteBtn ? "Delete Contact(s)" : title}
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button className="btn btn-danger" onClick={handleDeleteContact}>
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container p-5">
        <div className="display-end mt-4">
          <button
            className={`btn ${deleteBtn ? "btn-danger" : "btn-primary"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5">You have no contacts!</div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
