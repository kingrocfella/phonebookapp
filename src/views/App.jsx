import { useState } from "react";
import Modal from "../components/Modal";
import AddContact from "./addContact";
import ContactList from "./contactList";
import ContactDetails from "./contactDetails";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "../styles/App.scss";
import {
  BsPersonPlus,
  BsPersonDash,
  BsExclamationCircle,
} from "react-icons/bs";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState({});

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
    toast.success("Contact created!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
    toast.success("Contact successfully edited!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
    toast.success("Contact(s) deleted!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={
            deleteBtn
              ? "Delete Contact(s)"
              : modalType === CONSTANTS.contact
              ? "Add Contact"
              : title
          }
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button
                className="btn delete-button"
                onClick={handleDeleteContact}
              >
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container text-white">
        <div className="space-between mt-5">
          <h3>Phonebook App</h3>
          <button
            className={`btn datatest-2 ${deleteBtn ? "delete-button" : "primary-button"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? <BsPersonDash /> : <BsPersonPlus />} &nbsp;&nbsp;
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5 datatest-1">
              <BsExclamationCircle /> &nbsp; You have no contacts!
            </div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
