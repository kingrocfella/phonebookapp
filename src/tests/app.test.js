import React from "react";
import { shallow } from "enzyme";
import App from "../views/App";
import { findClassAtrr } from "./utils";

describe("<App /> on page load", () => {
  const container = shallow(<App />);

  it("should match the snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });

  it("Render the no contacts text element", () => {
    expect(findClassAtrr(container, "datatest-1").length).toEqual(1);
  });

  it("Render the add contact button", () => {
    expect(findClassAtrr(container, "datatest-2").length).toEqual(1);
  });
});