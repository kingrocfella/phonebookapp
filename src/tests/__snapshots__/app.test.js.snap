// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> on page load should match the snapshot 1`] = `"<div class=\\"Toastify\\"></div><div class=\\"container text-white\\"><div class=\\"space-between mt-5\\"><h3>Phonebook App</h3><button class=\\"btn datatest-2 primary-button\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 16 16\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" d=\\"M11 14s1 0 1-1-1-4-6-4-6 3-6 4 1 1 1 1h10zm-9.995-.944v-.002.002zM1.022 13h9.956a.274.274 0 00.014-.002l.008-.002c-.001-.246-.154-.986-.832-1.664C9.516 10.68 8.289 10 6 10c-2.29 0-3.516.68-4.168 1.332-.678.678-.83 1.418-.832 1.664a1.05 1.05 0 00.022.004zm9.974.056v-.002.002zM6 7a2 2 0 100-4 2 2 0 000 4zm3-2a3 3 0 11-6 0 3 3 0 016 0zm4.5 0a.5.5 0 01.5.5v2a.5.5 0 01-.5.5h-2a.5.5 0 010-1H13V5.5a.5.5 0 01.5-.5z\\" clip-rule=\\"evenodd\\"></path><path fill-rule=\\"evenodd\\" d=\\"M13 7.5a.5.5 0 01.5-.5h2a.5.5 0 010 1H14v1.5a.5.5 0 01-1 0v-2z\\" clip-rule=\\"evenodd\\"></path></svg>   Add Contact</button></div><div class=\\"mt-5\\"><div class=\\"text-center pt-5 datatest-1\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 16 16\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" d=\\"M8 15A7 7 0 108 1a7 7 0 000 14zm0 1A8 8 0 108 0a8 8 0 000 16z\\" clip-rule=\\"evenodd\\"></path><path d=\\"M7.002 11a1 1 0 112 0 1 1 0 01-2 0zM7.1 4.995a.905.905 0 111.8 0l-.35 3.507a.552.552 0 01-1.1 0L7.1 4.995z\\"></path></svg>   You have no contacts!</div></div></div>"`;
